### PR TITLE
Fix #13495: typechecker crash while typing objects 

### LINKED ISSUE #13495: [5.3] typechecker crash while typing objects
```
nojebar@PERVERSESHEAF:~/tmp$ cat bug.ml
class type gui =
  object
    method sub: 'b. 'b -> 'b
  end

class virtual local_sub =
  object
    method virtual sub: 'b. 'b -> 'b option
  end

class virtual ['a] compound_gui =
  object (_: #gui)
    constraint 'a = #local_sub
  end
nojebar@PERVERSESHEAF:~/tmp$ opam exec --switch 4.14.2 -- ocaml bug.ml
nojebar@PERVERSESHEAF:~/tmp$ opam exec --switch 5.1.0 -- ocaml bug.ml
nojebar@PERVERSESHEAF:~/tmp$ opam exec --switch 5.2.0 -- ocaml bug.ml
nojebar@PERVERSESHEAF:~/tmp$ opam exec --switch 5.3.0+trunk -- ocaml bug.ml
>> Fatal error: Ctype.unify_univar: univar not in scope
Fatal error: exception Misc.Fatal_error
```
(The example is extracted from a much larger codebase.)

### PR DESCRIPTION
**Pull request by garrigue**
The crash in issue #13495 was caused by the addition of a fatal error in `unify_univar`, which was triggered by calling `eqtype` on types containing univars.

The solution proposed consists of:
1) replacing the fatal error of PR #12920 by a normal failure, which just causes `eqtype` to return `false`, which is fine in the concerned place (it is an improvement to injectivity inferrence in Typedecl_variance).
2) weaken `mcomp`, so that is now ignores univars. Since `mcomp` is an approximation anyway, this is fine, and avoids potential soundness issues. Also added some explanations to `mcomp` to emphasize the approximation aspect.

### CODE REVIEW

File: Changes [hunk 1]
```diff
@@ -78,6 +78,9 @@ Working version
 - #13454: Output a correct trace of the C_CALLN bytecode.
   (Miod Vallat, review by Antonin Décimo)
 
+- #131495: Fix typechecker crash while typing objects
+  (Jacques Garrigue, report by Nicolás Ojeda Bär, review by ???)
```
Comment (nojb) [inline Changes]: The Changes entry should be in the 5.3.0 section.

File: typing/ctype.ml [hunk 2]
```diff
@@ -1978,7 +1978,7 @@ let rec unify_univar t1 t2 = function
           raise Cannot_unify_universal_variables
       end
   | [] ->
-     Misc.fatal_error "Ctype.unify_univar: univar not in scope"
+      raise Cannot_unify_universal_variables
```
Comment (nojb) [inline typing/ctype.ml]:
```suggestion
     raise Cannot_unify_universal_variables
```

File: typing/ctype.ml [hunk 3]
```diff
@@ -2330,12 +2330,21 @@ let rec expands_to_datatype env ty =
       end
   | _ -> false
 
-(* [mcomp] tests if two types are "compatible" -- i.e., if they could ever
-   unify.  (This is distinct from [eqtype], which checks if two types *are*
-   exactly the same.)  This is used to decide whether GADT cases are
-   unreachable.  It is broadly part of unification. *)
+(* [mcomp] tests if two types are "compatible" -- i.e., if there could
+   exist a witness of their equality. This is distinct from [eqtype],
+   which checks if two types *are*  exactly the same.)
```
Comment (nojb) [inline typing/ctype.ml]:
```suggestion
   which checks if two types *are*  exactly the same.
```

**Comment by gasche** [comment]
By "`mcomp` ignores univars", you mean that two univars are always considered potentially-compatible with the new approximation.

(Note: there may be room to refine this, because when both univars *are* in type_pairs we can decide whether they are compatible or incompatible, right? Univars in scope are distinct from all other univars in scope, except one of those in the same pair.)

Can this new approximation break code? Can univar includes in GADT-equated code? (Is it the case that univars only occur in type expressions where they are bound, and never free in term-level expressions as type variables and locally abstract types?)

@stedolan proposed the change to fatal_error on univars that are not in scope, maybe he has an opinion about turning this into "not equal" again.

**Comment by gasche** [comment]
Note: I think the worry of @stedolan was not that some parts of the compiler would check types in different univar contexts, but rather that we could check two types in the same univar contexts, but where the current implementation has erased a part of the `univar_pairs` global variable due to re-entrancy issues. In this case, if there is a bug of this form, then `eqtype` may return `false` when it should in fact return `true`, and this can cause unsoundness. From this perspective, the change proposed here is a regression, because it hides the issue under the carpet again.

We could probably have a different exception, `Unknown_univars`, for the case where a variable is completely absent from the current univar scope. (It's a bit subtle because currently we only reach this case when *both* univars are unknown, we would have to refactor the code if we wanted to detect that *either* univar is unknown. But maybe we don't need to, because the re-entrancy bug that we are worried about would only make both univars unknown.) But then if we let callers handle it, we may miss some of them or handle them incorrectly, this is no magic bullet.

**Comment by stedolan** [comment]
> @stedolan proposed the change to fatal_error on univars that are not in scope, maybe he has an opinion about turning this into "not equal" again.

Yes, my opinion is the following:

> Note: I think the worry of @stedolan was not that some parts of the compiler would check types in different univar contexts, but rather that we could check two types in the same univar contexts, but where the current implementation has erased a part of the `univar_pairs` global variable due to re-entrancy issues. In this case, if there is a bug of this form, then `eqtype` may return `false` when it should in fact return `true`, and this can cause unsoundness. From this perspective, the change proposed here is a regression, because it hides the issue under the carpet again.

Why do we have types with free univars anyway? Is there no surrounding `Tpoly` that binds them? Could we perhaps call `enter_poly` at the point where they are bound and then compare the types in the correct context, using a version of `Ctype.equal` that didn't reset `univar_pairs`?

**Comment by garrigue** [comment]
> Why do we have types with free univars anyway? Is there no surrounding `Tpoly` that binds them? Could we perhaps call `enter_poly` at the point where they are bound and then compare the types in the correct context, using a version of `Ctype.equal` that didn't reset `univar_pairs`?

Inside `Typedecl_variance` we are looking for nodes of a type that have equal subtrees, so that we can collect their variances. Those nodes being taken from anywhere in the type graph, they may contain different univars coming from different scopes, and there is no "correct" context. Of course, one could build an appropriate `univar_pairs` by collecting the univars on both sides, and allowing only reflexivity, but is it really worth it?

BTW, I don't see how `equal` returning `false` could cause an unsoundness: using such a negative information positively is unsound anyway, this is precisely the reason we need `mcomp`.

@gasche 
As for trying to be more precise in `mcomp`, I'm not sure it is worth the pain: for `mcomp`, correctness is much more important than precision. Except if you see useful concrete cases where this would make the exhaustiveness more incomplete.

**Comment by garrigue** [comment]
Thinking about "collecting all univars under a node", this is precisely what `Ctype.compute_univars` does, so performance is not going to be a problem. This is still a fair amount of extra code. Also, the algorithm in `compute_univars` would have to be made reentrant, because `Typedecl_variance.compute_variance_type` does some expansions.

**Comment by gasche** [comment]
I think that having a `Unpaired_univars` exception out of `unify_univar` would be a reasonable compromise: calls that expect to be in the same univar scope would handle it as a fatal error (as today in trunk), and calls that cannot assume anything about the univar scope handle it as an inequality (and hope for the best).

**Comment by stedolan** [comment]
> BTW, I don't see how equal returning false could cause an unsoundness: using such a negative information positively is unsound anyway, this is precisely the reason we need mcomp.

Yes, this makes sense.

I think the problem in #12910 was equal having a side-effect of clearing `univars`, which caused a call to `mcomp` to incorrectly think two types were distinguishable. (I think? I've forgotten the exact code path). That wouldn't occur just because equal spuriously returned false.

> I think that having a Unpaired_univars exception out of unify_univar would be a reasonable compromise

I like this suggestion: instead of equal returning false on cases where there are dangling univars, the places where this is known to be safe can catch the exception and consider it false, and every other place fails.

**Comment by garrigue** [comment]
I have tried to answer the requests.
For the different behaviors, I discriminate on the kind of exception expected, assuming a special role for `Equality`. It was not clear to me how to handle this clearly at call sites. Register a default handler with `Location.register_error_of_exn` ?

**Comment by Octachron** [comment]
Note that the weakening of `mcomp` will change some errors on recursive modules due to the use of `mcomp` to detect inconsistent type declarations between the implementation and the signature introduced in #1468. For instance, 
```ocaml
module rec M: sig
  type t = { r: 'a 'b. 'a -> ' b -> 'a }
   val f: t -> t
end = struct
  type t = { r : 'a. 'a -> 'a -> 'a }
   let f: t -> M.t = fun x -> x
 end
 ```
 now fails when checking the inclusion:
```
 Error: Signature mismatch:
       ...
       Type declarations do not match:
         type t = M.t = { r : 'a. 'a -> 'a -> 'a; }
       is not included in
         type t = { r : 'a 'b. 'a -> 'b -> 'a; }
```
rather than when checking `f` in the body
```ocaml
6 |    let f: t -> M.t = fun x -> x
                                  ^
Error: This expression has type t but an expression was expected of type M.t
```

because `mcomp t M.t` is true with this PR which makes `enrich_typedecl` enrich the type declaration of t to `type t = M.t = { r : 'a. 'a -> 'a -> 'a }`.

**Comment by garrigue** [comment]
@Octachron Do you mean that you would prefer going back to the original behavior of `mcomp` ?
I suppose this is fine, since we are now able to distinguish between the different causes of failure in `unify_univar`.
I had completely forgotten that `mcomp` was used for anything else than GADTs.

**Comment by garrigue** [comment]
I have rebased, and re-enabled the check of univars in `mcomp`.

**Comment by Octachron** [comment]
I am not sure if it was really needed to get back to the previous version of `mcomp`, but it makes the code easier to review since I was previously stuck pondering on the articulation of the changes in `mcomp` with the rest of the PR. The link is much clearer with the last version of the code.

**Review by Octachron — APPROVED** [review]
I agree that it is safe to conclude that in presence of dangling univar, two types are neither distinguishable (in `mcomp`) nor equal (in `eqtype`).

The current PR looks good to me.

File: typing/ctype.ml [hunk 4]
```diff
@@ -1969,13 +1971,19 @@ let rec unify_univar t1 t2 = function
...
-  try unify_univar t1 t2 univar_pairs
-  with Cannot_unify_universal_variables -> raise_unexplained_for tr_exn
+let unify_univar_for (type a) (tr_exn : a trace_exn) t1 t2 univar_pairs =
+  try unify_univar t1 t2 univar_pairs with
+  | Cannot_unify_universal_variables -> raise_unexplained_for tr_exn
+  | Out_of_scope_universal_variable ->
+      (* allow unscoped univars when checking for equality;
+         see Typedecl_variance (#13514) for instance *)
+      match tr_exn with
+      | Equality -> raise_unexplained_for tr_exn
```
Comment (Octachron) [inline typing/ctype.ml:1986]: Maybe we could add a comment that in a `Equality` context, a dangling univar can be interpreted as a non-fatal error?

**Comment by garrigue** [comment]
Merged into trunk and cherry-picked to 5.3.

**Pull request merged**